Ensure that resources are always initialized in all cases

zenspider opened the pull request:
It does NOT do what you think it does. It does NOT prevent an object
from coming back. All it does is have an uninitialized (or partially
initialized) object come back, which isn't helpful.

Signed-off-by: Ryan Davis <zenspider@chef.io>

zenspider committed:
Do not call return inside of an initialize method.

It does NOT do what you think it does. It does NOT prevent an object
from coming back. All it does is have an uninitialized (or partially
initialized) object come back, which isn't helpful.

Signed-off-by: Ryan Davis <zenspider@chef.io>

miah reviewed (approved):
I wonder about side effects on this one though

clintoncwolfe reviewed (changes requested):
I think it's actually intentional behavior with a bad comment - the intent really is not to affect the return value, but to just avoid further initializiation.

If that entire initialization block that follows were placed under an `unless`, that might be clearer, and not involve a suspicious return-from-initialize...

```
# Stop initializing if we are on an unsupported or test backend
unless supported || test_backend
  begin
    super(*args)
  rescue...
    ...
  end
end
```

clintoncwolfe reviewed (changes requested):
(double commented, whoops)

clintoncwolfe reviewed (approved):
If that's the intent - to always initialize resources - I'll accept, and I've updated the PR title for you to reflect the intent.

zenspider committed:
Fixes for 2 failed tests.

Rescue NotImplementedError and call failed_resource only if we haven't
already failed the support_check call.

Signed-off-by: Ryan Davis <zenspider@chef.io>
